README.md

```diff
@@ -1,48 +1,76 @@
 # fsctl
 
-Series of small native fd utils for manipulating file attributes and more
+Native utilities for file manipulation, including locking and hole punching.
 
 ```
 npm install fsctl
 ```
 
 ## Usage
 
+Write to a file using an exclusive lock:
+
 ``` js
-const { lock, unlock, sparse } = require('fsctl')
+const { open, write } = require('fs')
+const { lock, unlock } = require('fsctl')
+
+open('file.txt', 'a+', async (err, fd) => {
+  if (err) throw err
 
-// Can we lock the file using the fd?
-console.log(lock(fd))
+  await lock(fd, { exclusive: true })
 
-// Can we unlock it?
-console.log(unlock(fd))
+  write(fd, 'hello world', async (err) => {
+    if (err) throw err
 
-// Can we set the file as sparse?
-console.log(sparse(fd))
+    await unlock(fd)
+  })
+})
 ```
 
 ## API
 
-#### `bool = fsctl.lock(fd)`
+#### `await fsctl.lock(fd[, offset[, length]][, options])`
+
+Request a process level lock on a file, resolving when the lock is granted. If another process holds the lock, the lock will not be granted until the other process either exits or releases the lock.
+
+To lock only a portion of the file, `offset` and `length` may be passed. A `length` of `0` will request a lock from `offset` to the end of the file.
+
+Note that the lock is only advisory and there is nothing stopping another process from accessing the file by simply ignoring the lock.
+
+Options include:
+
+```js
+{
+  // If `true`, request an exclusive lock, i.e. a write lock, on the file. By
+  // default, a shared lock, i.e. a read lock, is requested.
+  // Be aware that an exclusive lock can only be granted to files that are
+  // writable! A request for an exclusive lock on a read-only file is ignored 
+  // and treated as a request for a shared lock.
+  exclusive: false
+}
+```
+
+#### `const granted = fsctl.tryLock(fd[, offset[, length]][, options])`
+
+Request a process level lock on a file, returning `true` if the lock was granted or `false` if another process currently holds the lock.
 
-Try to lock access to a file using a file descriptor.
-Returns true if the file could be locked, false if not.
+Options are the same as `fsctl.lock()`.
 
-Note that the lock is only advisory and there is nothing stopping someone from accessing the file by simply ignoring the lock.
+#### `fsctl.unlock(fd[, offset[, length]])`
 
-Works across processes as well.
+Release a process level lock on a file.
 
-#### `bool = fsctl.unlock(fd)`
+#### `await fsctl.punchHole(fd, offset, length)`
```

Comment:
is unlinkBlocks a better generic name for this? punchHole is obvs cool also 😎

Reply:
Also wondering if we should get rid of `setSparse()` and just do two `DeviceIoControl()` calls when hole punching on Windows. With that in place, would `unlinkBlocks(fd, index, count = 1)` make sense?

Reply:
That would obviously need to do a stat to get the block size first.

Reply:
Does the underlying syscall use blocks or bytes as offsets/length? We should just mimick that I guess if the same across OSs

Reply:
My suggestion would be to remove it and just do this on Windows:

```c
int
fsctl__punch_hole (uv_os_fd_t fd, uint64_t offset, size_t length) {
  FILE_ZERO_DATA_INFORMATION data = {
    .FileOffset = {
      .QuadPart = offset,
    },
    .BeyondFinalZero = {
      .QuadPart = offset + length,
    },
  };

  DWORD bytes;

  BOOL res = DeviceIoControl(
    fd,
    FSCTL_SET_SPARSE,
    NULL,
    0,
    NULL,
    0,
    &bytes, // Must be passed when lpOverlapped is NULL
    NULL
  );

  if (!res) return uv_translate_sys_error(GetLastError());

  res = DeviceIoControl(
    fd,
    FSCTL_SET_ZERO_DATA,
    &data,
    sizeof(data),
    NULL,
    0,
    &bytes, // Must be passed when lpOverlapped is NULL
    NULL
  );

  return res ? 0 : uv_translate_sys_error(GetLastError());
}
```

Reply:
Great point, I'm likely reading the Windows docs wrong! This is the bit that caught my eye:

> When you perform a write operation (with a function or operation other than FSCTL_SET_ZERO_DATA) whose data consists of nothing but zeros, zeros will be written to the disk for the entire length of the write. To zero out a range of the file and maintain sparseness, use FSCTL_SET_ZERO_DATA.

https://docs.microsoft.com/en-us/windows/win32/fileio/sparse-file-operations

I was therefore under the impression that sparseness wouldn't be maintained for normal writes like on non-Windows systems. Let me test that to make sure I got it wrong.

Regarding the hole punching APIs, they use byte offset and length on all systems that support them.

Reply:
Windows behaves like macOS and Linux on sparse writes so all is well and we definitely need `setSparse()` 👌

Reply:
Should we change the name?

Reply:
Let's keep it as is, I overthought it.